File: platform/api/services/scientific_analytics.py
"""Read-only, source-verified canonical scalar analytics for marked results."""
import hashlib
import json
import math
import statistics
from itertools import combinations
from typing import Literal
from pydantic import BaseModel, ConfigDict, Field, StrictFloat, StrictInt, model_validator
from sqlalchemy import select
from database import Job
from services.core_protein_scientific_contract import revision_for_job, validate_metric


class ClosedModel(BaseModel):
    model_config = ConfigDict(extra='forbid')


class MetricState(ClosedModel):
    state: Literal['ok', 'unavailable', 'invalid']
    value: StrictFloat | StrictInt | None
    reason_code: str | None

    @model_validator(mode='after')
    def coherent(self):
        if self.state == 'ok':
            if self.value is None or not math.isfinite(self.value) or self.reason_code is not None:
                raise ValueError('ok requires finite real value and null reason')
        elif self.value is not None or not self.reason_code or not self.reason_code.strip():
            raise ValueError('missing/invalid requires null value and reason')
        return self


class MetricDescriptor(ClosedModel):
    metric_id: str
    source: Literal['canonical_artifact']
    scope: str
    unit: str
    direction: Literal['higher', 'lower', 'none']
    producer_version: str
    derivation_version: str


class MetricSource(ClosedModel):
    artifact_sha256: str = Field(pattern=r'^[0-9a-f]{64}$')
    candidate_id: str = Field(min_length=1)
    document_id: str = Field(min_length=1)


class DistributionStatistics(ClosedModel):
    min: StrictFloat | StrictInt
    max: StrictFloat | StrictInt
    avg: StrictFloat | StrictInt
    median: StrictFloat | StrictInt
    std_dev: StrictFloat | StrictInt


class DistributionSummary(ClosedModel):
    observed_count: StrictInt = Field(ge=0)
    unavailable_count: StrictInt = Field(ge=0)
    invalid_count: StrictInt = Field(ge=0)
    descriptor: MetricDescriptor
    statistics: DistributionStatistics | None
    reason_code: str | None


class PairPoint(ClosedModel):
    id: str
    x: StrictFloat | StrictInt
    y: StrictFloat | StrictInt


class PairSummary(ClosedModel):
    x_metric: str
    y_metric: str
    pair_count: StrictInt = Field(ge=0)
    excluded_count: StrictInt = Field(ge=0)
    excluded_ids: list[str]
    points: list[PairPoint]
    correlation: MetricState


class ScientificCohort(ClosedModel):
    cohort_key: str
    design_ids: list[str]
    metrics: dict[str, DistributionSummary]
    pairs: dict[str, PairSummary]


# Owner-fixed native definitions also describe unreadable evidence without
# trusting a tampered compact block for its descriptor or supported metric list.
_BOLTZ = {'ptm': ('dimensionless', 'overall'), 'complex_plddt': ('fraction', 'complex')}


def metric_state(value):
    if value is None:
        return MetricState(state='unavailable', value=None, reason_code='not_reported')
    if type(value) not in (int, float) or not math.isfinite(value):
        return MetricState(state='invalid', value=None, reason_code='not_finite_real')
    return MetricState(state='ok', value=value, reason_code=None)


def projection(design, *, records=(), invalid_reason=None, model_id=None):
    """Project owner-verified records; never read legacy columns as canonical."""
    states, descriptors, sources = {}, {}, {}
    for record in records:
        record = validate_metric(record)
        key = record['metric_key']
        if key in states:
            raise ValueError('duplicate canonical metric')
        states[key] = MetricState(**{k:record[k] for k in ('state','value','reason_code')})
        descriptors[key] = MetricDescriptor(metric_id=key, source='canonical_artifact',
            scope=record['scope'], unit=record['unit'],
            direction={'higher_is_better':'higher','lower_is_better':'lower','neutral':'none'}[record['direction']],
            producer_version=record['producer_version'], derivation_version=record['derivation_version'])
        sources[key] = MetricSource.model_validate(record['source']).model_dump()
    if invalid_reason:
        expected = _BOLTZ if model_id in ('boltz', 'boltz2') else {}
        if model_id in ('esmfold2', 'esmfold2_experimental'):
            from services.esmfold2_scientific_consumer import DESCRIPTORS
            expected = {d['metric_key']: (d['unit'], d['scope']) for d in DESCRIPTORS}
        if model_id in ('boltzgen', 'boltzgen_child'):
            # Fixed native semantics only; no artifact or alignment subtype is
            # claimed when the publication cannot be independently verified.
            expected = {
                'design_ptm': ('fraction', 'native_design_chain_tokens'),
                'affinity_probability': ('fraction', 'native_affinity_binary1_complex'),
                'filter_rmsd': ('angstrom', 'native_filter_complex_alignment'),
            }
        for key, (unit, scope) in expected.items():
            states[key] = MetricState(state='invalid', value=None, reason_code=invalid_reason)
            descriptors[key] = MetricDescriptor(metric_id=key, source='canonical_artifact', scope=scope,
                unit=unit, direction='lower' if key == 'filter_rmsd' else 'higher',
                producer_version='unverified', derivation_version='unverified')
            sources[key] = None
    signature = json.dumps({k:d.model_dump() for k,d in sorted(descriptors.items())}, sort_keys=True, separators=(',',':'))
    key = hashlib.sha256(signature.encode()).hexdigest()
    return dict(contract_revision=1, source_job_id=design.job_id,
        cohort_key=f'v1:{key}:{design.job_id}', metric_states=states,
        metric_descriptors=descriptors, metric_sources=sources,
        metrics={k:s.value for k,s in states.items() if s.state == 'ok'})


async def persisted_projection(design, session):
    """Keep owning-Job lookup and publication verification in the same read."""
    with session.no_autoflush:
        job = await session.get(Job, design.job_id)
        if job is None or revision_for_job(job) != 1:
            raise ValueError('canonical projection requires marked owning Job')
        try:
            if job.model_id in ('boltz', 'boltz2'):
                from services.boltz_scientific_consumer import verified_boltz_design
                selected = await verified_boltz_design(design, session)
                return projection(design, records=selected['block']['metrics'])
            if job.model_id in ('esmfold2', 'esmfold2_experimental'):
                from services.esmfold2_scientific_consumer import verified_esmfold2_design
                selected = await verified_esmfold2_design(design, session)
                return projection(design, records=selected['block']['metrics'])
            if job.model_id in ('boltzgen', 'boltzgen_child'):
                from services.boltzgen_candidate_publication import verified_boltzgen_design
                selected = await verified_boltzgen_design(session, design)
                return projection(design, records=selected['block']['metrics'].values())
            # Other model adapters must independently validate native source
            # bytes before contributing canonical scalar records here.
            return projection(design, invalid_reason='missing_canonical_publication', model_id=job.model_id)
        except (ValueError, TypeError, KeyError, IndexError, OSError, RuntimeError):
            return projection(design, invalid_reason='invalid_canonical_publication', model_id=job.model_id)


def summarize(rows):
    """Rows are (Design ID, verified projection) pairs from one descriptor cohort."""
    ids = [identity for identity, _ in rows]
    if len(set(ids)) != len(ids):
        raise ValueError('duplicate analytics candidate')
    metrics = {}
    for key in sorted({key for _, p in rows for key in p['metric_states']}):
        descriptors = [p['metric_descriptors'][key] for _,p in rows]
        if any(d != descriptors[0] for d in descriptors):
            raise ValueError('incompatible metric descriptors')
        states = [p['metric_states'][key] for _,p in rows]
        values = [s.value for s in states if s.state == 'ok']
        metrics[key] = dict(observed_count=len(values),
            unavailable_count=sum(s.state == 'unavailable' for s in states),
            invalid_count=sum(s.state == 'invalid' for s in states), descriptor=descriptors[0],
            statistics=dict(min=min(values), max=max(values), avg=statistics.mean(values),
                median=statistics.median(values), std_dev=statistics.pstdev(values)) if values else None,
            reason_code=None if values else 'no_observed_values')
    pairs = {}
    for x, y in combinations(sorted(metrics), 2):
        points = [dict(id=id, x=p['metrics'][x], y=p['metrics'][y])
            for id,p in rows if x in p['metrics'] and y in p['metrics']]
        included = {p['id'] for p in points}
        correlation = MetricState(state='unavailable', value=None, reason_code='insufficient_pairs')
        if len(points) >= 3:
            try:
                correlation = metric_state(statistics.correlation([p['x'] for p in points], [p['y'] for p in points]))
            except statistics.StatisticsError:
                correlation = MetricState(state='unavailable', value=None, reason_code='constant_metric')
        pairs[f'{x}_vs_{y}'] = dict(x_metric=x, y_metric=y, points=points, pair_count=len(points),
            excluded_count=len(rows)-len(points), excluded_ids=[i for i in ids if i not in included], correlation=correlation)
    return dict(metrics=metrics, pairs=pairs)


async def partition(designs, owners, session, projections=None):
    legacy, grouped = [], {}
    for design in designs:
        if revision_for_job(owners.get(design.job_id)) != 1:
            legacy.append(design)
        else:
            value = projections[design.id] if projections is not None else await persisted_projection(design, session)
            grouped.setdefault(value['cohort_key'], []).append((design.id, value))
    return legacy, [ScientificCohort(cohort_key=key, design_ids=[id for id,_ in rows], **summarize(rows))
        for key, rows in grouped.items()]


async def owning_jobs(session, designs):
    ids = {d.job_id for d in designs}
    if not ids:
        return {}
    rows = (await session.execute(select(Job).where(Job.id.in_(ids)))).scalars().all()
    return {job.id: job for job in rows}

File: platform/frontend/src/lib/scientificAnalytics.ts
export type MetricState = {state:'ok'; value:number; reason_code:null} | {state:'unavailable'|'invalid'; value:null; reason_code:string};
export interface MetricDescriptor {metric_id:string; source:'canonical_artifact'; scope:string; unit:string; direction:'higher'|'lower'|'none'; producer_version:string; derivation_version:string}
export interface MetricSource {artifact_sha256:string; candidate_id:string; document_id:string}
export interface ScientificPoint {
    id:string; name:string; contract_revision:1; source_job_id:string; cohort_key:string;
    metrics:Record<string,number>; metric_states:Record<string,MetricState>;
    metric_descriptors:Record<string,MetricDescriptor>; metric_sources:Record<string,MetricSource|null>;
}
export interface ScientificCohort {
    cohort_key:string; design_ids:string[];
    metrics:Record<string,{observed_count:number; unavailable_count:number; invalid_count:number; descriptor:MetricDescriptor; statistics:Record<'min'|'max'|'avg'|'median'|'std_dev',number>|null; reason_code:string|null}>;
    pairs:Record<string,{x_metric:string;y_metric:string;pair_count:number;excluded_count:number;excluded_ids:string[];points:{id:string;x:number;y:number}[];correlation:MetricState}>;
}
const allowed=new Set(['design_ptm','affinity_probability','filter_rmsd','complex_plddt','plddt','plddt_mean','iptm','plddt_overall','plddt_binder','pae_overall','pae_interaction','rmsd_overall','rmsd_binder','ptm','conf_score','rog','mpnn_score','fampnn_psce','ligand_iptm','affinity_score','binder_probability','frustration_pct_high','maturation_interface_score','maturation_rmsd','maturation_delta_interface','maturation_selected_interface_score','maturation_selected_rmsd','maturation_nonselected_rmsd','ppiflow_objective_score','ppiflow_primary_loop_rmsd']);
function requireThat(condition:unknown):asserts condition {if(!condition) throw new Error('Invalid scientific analytics contract');}
function object(value:unknown):Record<string,unknown> {requireThat(value!==null&&typeof value==='object'&&!Array.isArray(value));return value as Record<string,unknown>;}
function keys(value:Record<string,unknown>,expected:string[]) {requireThat(Object.keys(value).sort().join('|')===[...expected].sort().join('|'));}
function text(value:unknown):asserts value is string {requireThat(typeof value==='string'&&value.trim().length>0);}
function finite(value:unknown):asserts value is number {requireThat(typeof value==='number'&&Number.isFinite(value));}
function count(value:unknown):asserts value is number {finite(value);requireThat(Number.isSafeInteger(value)&&value>=0);}
function identities(value:unknown):string[] {requireThat(Array.isArray(value));value.forEach(text);requireThat(new Set(value).size===value.length);return value;}
function sameDescriptor(a:MetricDescriptor,b:MetricDescriptor) {return Object.keys(a).every(k=>a[k as keyof MetricDescriptor]===b[k as keyof MetricDescriptor]);}
export function parseMetricState(value:unknown):MetricState {
    const row=object(value);keys(row,['state','value','reason_code']);
    if(row.state==='ok'){finite(row.value);requireThat(row.reason_code===null);}
    else{requireThat(row.state==='unavailable'||row.state==='invalid');requireThat(row.value===null);text(row.reason_code);}
    return value as MetricState;
}
function descriptor(value:unknown,key:string):MetricDescriptor {
    const row=object(value);keys(row,['metric_id','source','scope','unit','direction','producer_version','derivation_version']);
    requireThat(allowed.has(key)&&row.metric_id===key&&row.source==='canonical_artifact');
    text(row.scope);text(row.unit);text(row.producer_version);text(row.derivation_version);
    requireThat(row.direction==='higher'||row.direction==='lower'||row.direction==='none');
    return value as MetricDescriptor;
}
export function parseScientificPoint(value:unknown):ScientificPoint {
    const row=object(value);keys(row,['id','name','contract_revision','source_job_id','cohort_key','metrics','metric_states','metric_descriptors','metric_sources']);
    requireThat(row.contract_revision===1);text(row.id);text(row.name);text(row.source_job_id);text(row.cohort_key);
    requireThat(row.cohort_key.startsWith('v1:')&&row.cohort_key.endsWith(`:${row.source_job_id}`));
    const states=object(row.metric_states),descriptors=object(row.metric_descriptors),metrics=object(row.metrics),sources=object(row.metric_sources);
    keys(descriptors,Object.keys(states));keys(sources,Object.keys(states));
    const observed:string[]=[];let binding:string|undefined;
    for(const [key,value] of Object.entries(states)) {
        const d=descriptor(descriptors[key],key),state=parseMetricState(value);
        if(state.state==='ok'){observed.push(key);requireThat(metrics[key]===state.value);requireThat(d.producer_version!=='unverified'&&d.derivation_version!=='unverified');}
        if(sources[key]===null){requireThat(state.state!=='ok');continue;}
        const source=object(sources[key]);keys(source,['artifact_sha256','candidate_id','document_id']);
        text(source.artifact_sha256);requireThat(/^[a-f0-9]{64}$/.test(source.artifact_sha256));text(source.candidate_id);text(source.document_id);
        const current=JSON.stringify([source.candidate_id,source.document_id]);requireThat(binding===undefined||binding===current);binding=current;
    }
    keys(metrics,observed);return value as ScientificPoint;
}
export function parseMetricPoints(value:unknown):Array<ScientificPoint|{id:string;name:string;metrics:Record<string,number>;contract_revision?:null}> {
    requireThat(Array.isArray(value));const ids=new Set<string>();
    for(const item of value){const row=object(item);text(row.id);requireThat(!ids.has(row.id));ids.add(row.id);
        if(row.contract_revision!=null)parseScientificPoint(row);
        else{
            for(const key of ['source_job_id','cohort_key','metric_states','metric_descriptors','metric_sources'])requireThat(row[key]==null);
            text(row.name);Object.values(object(row.metrics)).forEach(finite);
        }}
    return value as ReturnType<typeof parseMetricPoints>;
}
export function parseScientificCohorts(value:unknown):ScientificCohort[] {
    requireThat(Array.isArray(value));const cohortKeys=new Set<string>();const allIds=new Set<string>();
    for(const item of value) {
        const row=object(item);keys(row,['cohort_key','design_ids','metrics','pairs']);text(row.cohort_key);
        requireThat(!cohortKeys.has(row.cohort_key));cohortKeys.add(row.cohort_key);
        const ordered=identities(row.design_ids),ids=new Set(ordered);for(const id of ids){requireThat(!allIds.has(id));allIds.add(id);}
        const metrics=object(row.metrics);
        for(const [key,value] of Object.entries(metrics)) {
            const m=object(value);keys(m,['observed_count','unavailable_count','invalid_count','descriptor','statistics','reason_code']);
            count(m.observed_count);count(m.unavailable_count);count(m.invalid_count);requireThat(m.observed_count+m.unavailable_count+m.invalid_count===ids.size);descriptor(m.descriptor,key);
            if(m.observed_count===0){requireThat(m.statistics===null);text(m.reason_code);}
            else{const s=object(m.statistics);keys(s,['min','max','avg','median','std_dev']);Object.values(s).forEach(finite);requireThat(m.reason_code===null);requireThat((s.min as number)<=(s.avg as number)&&(s.avg as number)<=(s.max as number)&&(s.min as number)<=(s.median as number)&&(s.median as number)<=(s.max as number)&&(s.std_dev as number)>=0);}
        }
        for(const [key,value] of Object.entries(object(row.pairs))) {
            const p=object(value);keys(p,['x_metric','y_metric','pair_count','excluded_count','excluded_ids','points','correlation']);text(p.x_metric);text(p.y_metric);
            requireThat(p.x_metric!==p.y_metric&&key===`${p.x_metric}_vs_${p.y_metric}`&&Object.hasOwn(metrics,p.x_metric)&&Object.hasOwn(metrics,p.y_metric));
            count(p.pair_count);count(p.excluded_count);const excluded=identities(p.excluded_ids);requireThat(excluded.length===p.excluded_count);
            requireThat(Array.isArray(p.points)&&p.points.length===p.pair_count&&p.pair_count+p.excluded_count===ids.size);
            const seen=new Set<string>();
            for(const value of p.points){const point=object(value);keys(point,['id','x','y']);text(point.id);requireThat(ids.has(point.id)&&!seen.has(point.id));seen.add(point.id);finite(point.x);finite(point.y);}
            for(const id of excluded){requireThat(ids.has(id)&&!seen.has(id));seen.add(id);}
            const state=parseMetricState(p.correlation);requireThat(state.state!=='ok'||(p.pair_count>=3&&state.value>=-1&&state.value<=1));
        }
    }
    return value as ScientificCohort[];
}
export function validateScientificEnvelope<T>(value:T):T {
    const row=object(value);
    const cohorts=row.scientific_cohorts===undefined?[]:parseScientificCohorts(row.scientific_cohorts);
    const points=row.points===undefined?[]:parseMetricPoints(row.points).filter((p):p is ScientificPoint=>p.contract_revision===1);
    const byId=new Map(points.map(p=>[p.id,p]));
    if(row.points!==undefined)for(const cohort of cohorts) {
        const members=cohort.design_ids.map(id=>{const p=byId.get(id);requireThat(p&&p.cohort_key===cohort.cohort_key);return p;});
        for(const p of members){keys(p.metric_descriptors,Object.keys(cohort.metrics));for(const [key,m] of Object.entries(cohort.metrics))requireThat(sameDescriptor(m.descriptor,p.metric_descriptors[key]));}
        for(const pair of Object.values(cohort.pairs)) {
            const expected=members.filter(p=>p.metric_states[pair.x_metric].state==='ok'&&p.metric_states[pair.y_metric].state==='ok');
            requireThat(expected.length===pair.points.length);
            expected.forEach((p,i)=>{const point=pair.points[i];requireThat(point.id===p.id&&point.x===p.metrics[pair.x_metric]&&point.y===p.metrics[pair.y_metric]);});
            const excluded=members.filter(p=>!expected.includes(p)).map(p=>p.id);
            requireThat(JSON.stringify(excluded)===JSON.stringify(pair.excluded_ids));
        }
    }
    return value;
}
